### PR TITLE
Dev erosion

### PR DESCRIPTION
**Pull request by apyskir**
Preparing eroded masks. Eroding given % of original mask, currently: [10, 20, 30]. Heuristic used to select proper selem_size.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -53,20 +53,28 @@ def prepare_metadata(train_data, valid_data, test_data, public_paths):
 @action.command()
 @click.option('-d', '--dev_mode', help='if true only a small sample of data will be used', is_flag=True, required=False)
 def prepare_masks(dev_mode):
-    if params.erode_selem_size > 0:
-        erode = params.erode_selem_size
-        target_dir = params.masks_overlayed_eroded_dir
+    erode = eval(params.erosion_percents)
+    if erode != [0]:
```
Comment (jakubczakon) [inline main.py]: @apyskir I'd prefer erode[0] = 0 but I don't exactly know (or can show example) why

File: preparation.py [hunk 2]
```diff
@@ -83,6 +82,37 @@ def add_dropped_objects(original, processed):
...
-        if np.any(np.where(~(labeled==i) & processed)):
+        if np.any(np.where(~(labeled == i) & processed)):
             reconstructed += (labeled == i)
     return reconstructed.astype('uint8')
+
+
+def get_selem_size(mask, percent):
+    mask_area = np.sum(mask)
+    radius = np.sqrt(mask_area)
+    result = max(2, int(radius * percent / 200))
```
[the remark this answers is not in the document]
Comment (apyskir) [reply preparation.py]: @jakubczakon No, `selem_size` is always 2 or more. 1 or less makes no sense, right?
I don't know how much faster, but when your proper `selem_size` is, let's say, 10, then this function should do it in 2 or 3 iterations.

File: preparation.py [hunk 3]
```diff
@@ -55,15 +55,14 @@ def overlay_masks_from_annotations(annotations, image_size):
...
+def overlay_eroded_masks_from_annotations(annotations, image_size, area_percent):
     mask = np.zeros(image_size)
-    selem = rectangle(selem_size, selem_size)
     for ann in annotations:
         rle = cocomask.frPyObjects(ann['segmentation'], image_size[0], image_size[1])
         m = cocomask.decode(rle)
         m = m.reshape(image_size)
-        m = binary_erosion(m, selem=selem)
-        mask += m
+        m_eroded = get_eroded_mask(m, area_percent)
```
[the remark this answers is not in the document]
Comment (apyskir) [reply preparation.py]: Ok.

File: preparation.py [hunk 4]
```diff
@@ -55,15 +55,14 @@ def overlay_masks_from_annotations(annotations, image_size):
     return np.where(mask > 0, 1, 0).astype('uint8')
 
 
-def overlay_eroded_masks_from_annotations(annotations, image_size, selem_size):
+def overlay_eroded_masks_from_annotations(annotations, image_size, area_percent):
```
[the remark this answers is not in the document]
Comment (apyskir) [reply preparation.py]: Ok

File: neptune.yaml [hunk 5]
```diff
@@ -83,7 +83,7 @@ parameters:
 # Postprocessing
   threshold: 0.5
   min_nuclei_size: 20
-  erode_selem_size: 5
+  erosion_percents: '[10,20,30]'
```
[the remark this answers is not in the document]
Comment (apyskir) [reply neptune.yaml]: Ok

File: preparation.py [hunk 6]
```diff
@@ -83,6 +82,37 @@ def add_dropped_objects(original, processed):
...
+    while abs(diff) > 5 and iterations < 4:
+        iterations += 1
+        selem_size = get_selem_size(mask, new_percent)
+        selem = rectangle(selem_size, selem_size)
+        mask_eroded = binary_erosion(mask, selem=selem)
+        mask_area = np.sum(mask)
+        mask_eroded_area = np.sum(mask_eroded)
+        percent_obtained = 100 * (1 - mask_eroded_area / mask_area)
+        diff = percent - percent_obtained
+        new_percent = new_percent + diff
```
[the remark this answers is not in the document]
Comment (apyskir) [reply preparation.py]: OK

File: preparation.py [hunk 7]
```diff
@@ -83,6 +82,37 @@ def add_dropped_objects(original, processed):
...
+    result = max(2, int(radius * percent / 200))
+    return result
+
+
+def get_eroded_mask(mask, percent):
+    diff = 100
+    new_percent = percent
+    iterations = 0
+    while abs(diff) > 5 and iterations < 4:
+        iterations += 1
```
[the remark this answers is not in the document]
Comment (apyskir) [reply preparation.py]: @jakubczakon As you wish, it will be at the end

File: main.py (hunk 1 above)
Comment (apyskir) [reply main.py]: @jakubczakon You can imagine, that you want to prepare masks, that are not eroded, that are 10% and 20% eroded, so you pass a list [0,10,20]. Then your condition will prepare only not eroded masks, what is not what you want. I'll leave it as it is, ok?

File: preparation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jakubczakon) [reply preparation.py]: @apyskir you are right sorry

**Pull request merged**